### ISSUE TITLE
Unhandled `nil` when partitioning is in use

### ISSUE THREAD

**Issue by up2jj** [oban-bg/oban#1168]
### Environment

- Oban v2.17.10, Oban.Pro v1.4.9
- Elixir 1.16.2 (compiled with Erlang/OTP 26)

### Current Behavior

Previously, we had a queue declared in following way:
```elixir
my_queue: [
      local_limit: 1,
      global_limit: [allowed: 1, partition: [fields: [:args], keys: [:company_id]]]
```
Recently we have decided to:
1. Reconfigure queue:
```elixir
my_queue: [
      local_limit: 5,
      global_limit: [allowed: 3, partition: [:worker, args: :company_id]]
    ]
```
2. add a new queue:
```elixir
my_queue_2: [
      local_limit: 5,
      rate_limit: [
        allowed: 2000,
        period: {30, :minutes},
        partition: [:worker, args: :company_id]
```
3. reconfigure workers to use a new queue: `my_queue_2`

We have two kind of jobs currently available:
1. jobs with `company_id` key included in args,
2. jobs without that particular key.

Some of them were enqueued before configuration change.

During deployment we observed this error:
```elixir
GenServer {Oban.Registry, {Oban, {:producer, "my_queue"}}} terminating
** (ArgumentError) comparison with nil is forbidden as it is unsafe. If you want to check if a value is nil, use is_nil/1 instead
    (ecto 3.11.2) lib/ecto/query/builder.ex:1069: Ecto.Query.Builder.not_nil!/1
    (oban_pro 1.4.9) lib/oban/pro/engines/smart.ex:1071: anonymous fn/5 in Oban.Pro.Engines.Smart.demands_to_conditions/2
    (ecto 3.11.2) lib/ecto/query/builder/dynamic.ex:79: Ecto.Query.Builder.Dynamic.expand/3
    (ecto 3.11.2) lib/ecto/query/builder/dynamic.ex:49: Ecto.Query.Builder.Dynamic.fully_expand/2
    (ecto 3.11.2) lib/ecto/query/builder/filter.ex:133: Ecto.Query.Builder.Filter.filter!/7
    (oban_pro 1.4.9) lib/oban/pro/engines/smart.ex:986: Oban.Pro.Engines.Smart.fetch_subquery/4
    (oban_pro 1.4.9) lib/oban/pro/engines/smart.ex:886: Oban.Pro.Engines.Smart.fetch_jobs/2
    (ecto 3.11.2) lib/ecto/multi.ex:883: Ecto.Multi.apply_operation/5
    (elixir 1.16.2) lib/enum.ex:2528: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto 3.11.2) lib/ecto/multi.ex:856: anonymous fn/5 in Ecto.Multi.apply_operations/5
    (ecto_sql 3.11.3) lib/ecto/adapters/sql.ex:1358: anonymous fn/3 in Ecto.Adapters.SQL.checkout_or_transaction/4
    (db_connection 2.7.0) lib/db_connection.ex:1756: DBConnection.run_transaction/4
    (ecto 3.11.2) lib/ecto/repo/transaction.ex:18: Ecto.Repo.Transaction.transaction/4
    (oban_pro 1.4.9) lib/oban/pro/engines/smart.ex:536: anonymous fn/4 in Oban.Pro.Engines.Smart.fetch_jobs/3
    (oban_pro 1.4.9) lib/oban/pro/engines/smart.ex:1420: Oban.Pro.Engines.Smart.with_retry/3
    (oban 2.17.10) lib/oban/engine.ex:223: anonymous fn/4 in Oban.Engine.fetch_jobs/3
    (oban 2.17.10) lib/oban/engine.ex:319: anonymous fn/3 in Oban.Engine.with_span/4
    (telemetry 1.2.1) /app/deps/telemetry/src/telemetry.erl:321: :telemetry.span/3
    (oban 2.17.10) lib/oban/queue/producer.ex:245: Oban.Queue.Producer.start_jobs/1
    (oban 2.17.10) lib/oban/queue/producer.ex:236: anonymous fn/2 in Oban.Queue.Producer.dispatch/1
Last message: :dispatch 
```

The only way to get the system working again quickly was to pause the queue.

**Comment by sorentwo** [comment]
How are the queues defined? Was the existing queue reconfigured while the system was running, or as part of a deploy?

**Comment by up2jj** [comment]
We had to restore previous configuration of `my_queue` to maintain system operability. Current configuration which just work is:
```elixir
iex(node@node)1> Oban.check_queue(queue: :my_queue)
%{
  name: "Oban",
  node: "node@node",
  running: [],
  queue: "my_queue",
  started_at: ~U[2024-10-30 14:22:48.081766Z],
  uuid: "0192ddce-c311-7193-b3d7-1a21b2a2464f",
  updated_at: ~U[2024-10-31 06:53:41.989277Z],
  rate_limit: nil,
  paused: false,
  local_limit: 5,
  global_limit: %Oban.Pro.Producer.Meta.GlobalLimit{
    allowed: 3,
    tracked: %{},
    partition: %Oban.Pro.Producer.Meta.GlobalLimit.Partition{
      fields: [:args],
      keys: ["company_id"]
    }
  },
  retry_attempts: 5,
  retry_backoff: 1000
}

iex(node@node)2> Oban.check_queue(queue: :my_queue_2)
%{
  name: "Oban",
  node: "node@node",
  running: [],
  queue: "my_queue_2",
  started_at: ~U[2024-10-30 14:22:48.082374Z],
  uuid: "0192ddce-c312-777f-9778-c6f95ac34a9a",
  updated_at: ~U[2024-10-31 06:54:54.158268Z],
  rate_limit: %Oban.Pro.Producer.Meta.RateLimit{
    allowed: 2000,
    period: 1800,
    window_time: 1730356270,
    windows: [%{"curr_count" => 9, "prev_count" => 16}],
    partition: %Oban.Pro.Producer.Meta.RateLimit.Partition{
      fields: [:args, :worker],
      keys: ["company_id"]
    }
  },
  paused: false,
  local_limit: 5,
  global_limit: nil,
  retry_attempts: 5,
  retry_backoff: 1000
}
```

Queues were not reconfigured in runtime. We just edited config files, as a part of standard deployment process.

**Comment by sorentwo** [comment]
This is an edge case caused by the change from just `args` to both `worker` and `args`. It's fixed on main for inclusion in v1.5
